chronosphere/resource_resource_pools_config.go

```diff
@@ -36,112 +32,122 @@ import (
-			Allocation: expandAllocation(m.DefaultPool.Allocation),
-			Priorities: expandPriorities(m.DefaultPool.Priorities),
-		},
-		Pool: expandPools(m.Pools),
-	}
+func ResourcePoolsConfigFromModel(m *apimodels.Configv1ResourcePools) (*intschema.ResourcePoolsConfig, error) {
+	return (resourcePoolsConfigConverter{}).fromModel(m)
 }
 
 func resourceResourcePoolsConfig() *schema.Resource {
+	r := newGenericResource[
+		*models.Configv1ResourcePools,
+		intschema.ResourcePoolsConfig,
+		*intschema.ResourcePoolsConfig,
+	](
+		"resource_pools_config",
+		resourcePoolsConfigConverter{},
+		generatedResourcePools{},
+	)
+
 	return &schema.Resource{
+		CreateContext: r.CreateContext,
+		ReadContext:   r.ReadContext,
+		UpdateContext: r.UpdateContext,
+		DeleteContext: r.DeleteContext,
 		Schema:        tfschema.ResourcePoolsConfig,
+		CustomizeDiff: r.ValidateDryRun(&ResourcePoolsConfigDryRunCount),
 		SchemaVersion: 1,
 		Description:   "Shared admin config controlling quota usage in Chronosphere",
-		CreateContext: resourceResourcePoolsConfigCreate,
-		ReadContext:   resourceResourcePoolsConfigRead,
-		UpdateContext: resourceResourcePoolsConfigUpdate,
-		DeleteContext: resourceResourcePoolsConfigDelete,
-		CustomizeDiff: resourceResourcePoolCustomizeDiff,
 		Importer: &schema.ResourceImporter{
 			StateContext: schema.ImportStatePassthroughContext,
 		},
 	}
 }
 
-func resourceResourcePoolsConfigCreate(
-	ctx context.Context, d *schema.ResourceData, meta any,
-) diag.Diagnostics {
-	ctx = tfresource.NewContext(ctx, "resource_pools_config")
-	cli := getConfigClient(meta)
+// ResourcePoolsConfigDryRunCount tracks how many times dry run is run during validation for testing.
+var ResourcePoolsConfigDryRunCount atomic.Int64
 
-	poolsConfig, err := buildResourcePoolsConfig(d)
-	if err != nil {
-		return diag.Errorf("could not build resource pools config: %v", err)
-	}
+type resourcePoolsConfigConverter struct{}
 
-	req := &resource_pools.CreateResourcePoolsParams{
-		Body: &apimodels.Configv1CreateResourcePoolsRequest{
-			ResourcePools: poolsConfig,
-		},
-		Context: ctx,
-	}
-	if _, err = cli.ResourcePools.CreateResourcePools(req); err != nil {
-		return diag.Errorf("could not create resource pool config: %v", err)
+func (resourcePoolsConfigConverter) toModel(
+	r *intschema.ResourcePoolsConfig,
+) (*models.Configv1ResourcePools, error) {
+	if len(r.Pool) > 0 && len(r.Pools) > 0 {
+		return nil, fmt.Errorf("cannot set both pool and pools")
 	}
 
-	d.SetId(ResourcePoolsConfigID)
-
-	return nil
-}
-
-func resourceResourcePoolsConfigRead(
-	ctx context.Context, d *schema.ResourceData, meta any,
-) diag.Diagnostics {
-	ctx = tfresource.NewContext(ctx, "resource_pools_config")
-	cli := getConfigClient(meta)
+	// The resource has equivalent "pool" and deprecated "pools" lists.
+	// Only one is set at any given time and the server does not distinguish the different lists.
+	resourcePools := r.Pool
+	if len(r.Pools) > 0 {
+		resourcePools = r.Pools
```

Comment:
this implies that we favor the deprecated `r.Pools` value if they are both set. the comment seems to imply that is impossible so its probably fine but it does feel a bit backward

Reply:
This was a move of existing code, and agree it does't make a functional difference since only one can be set (check right above this ensures only one of them is non-empty). However, for clarity, it makes sense to show the `Pool` as preferred so updated.